### PR TITLE
Lists: Undo Exceptions Mark III

### PR DESCRIPTION
**Pull request by jleandroperez**
### Fix
In this PR we're reimplementing the `Insert / Remove List` logic:

A.  We're not inserting `- [ ]` and (later) replacing it with an Attachment anymore
B.  **SPTextAttachment** is now either inserted or removed directly
C.  A million new Unit Tests have been added

**Why:** Fixes Undo Exceptions triggered after Undoing either an Insert or Remove List OP

Eric!!! Apologies about bugging you so much!!! (and thanks a million in advance!!)
cc @aerych

Ref. #467

---

### Scenario A: Lists on Empty Documents
1. Add a new note
2. Press Option + Command + C to insert a Checklist

- [x] Verify the Item Marker shows up OnScreen
- [x] Verify that pressing Option + Command + C nukes the list

### Scenario B: Undo
1. Add a new note
2. Write a title and press **return**
3. Press **Option + Command + C** to insert a Checklist
4. Click back into the title and add a character
5. Click back on the list item
6. Press **Option + Command + C** to remove the Checklist

- [x] Verify that pressing CMD + Z brings back the list

### Scenario C: Preserving Cursor's Location
1. Add a new note
2. Add a few lines (with a word in each)
3. Click anywhere in the note
4. Press **Option + Command + C** to insert a Checklist

- [x] Verify the cursor remains in the same spot, relative to the word where you've clicked
- [x] Verify that pressing **Option + Command + C** removes the checklist and preserves the cursor's (relative) location

### Scenario D: Toggling Lists with Selection
1. Insert the following (sample) text in a new note:
    ```
    Line 1: No Leading Spaces
    Line 2: No Leading Spaces
            Line 3: This line contains 8 whitespaces
        Line 4: And this one contains only 4
    ```
2. Make sure that the last line is empty
3. Select everything
4. Press Option + Command + C to insert a Checklist

- [x] Verify every line gets a List Marker
- [x] Verify that the last (empty) line does not get a Marker
- [x] Verify the leading spaces are preserved!
- [x] Verify that selecting everything (again) and pressing **Option + Command + C** removes the list
- [x] Verify that pressing CMD + Z brings back the list

### Release
These changes do not require release notes.

### CODE REVIEW

File: SimplenoteTests/NSStringSimplenoteTests.swift [hunk 1]
```diff
@@ -104,4 +174,85 @@ class NSStringSimplenoteTests: XCTestCase {
...
+            .attachmentString + .space + "L3" + .newline,
+            "L4\n",
+            "L5"
+        ]
+        let expected = "L1\nL2\n\nL3\nL4\nL5"
+
+        XCTAssertEqual(sample.joined().removingListMarkers, expected)
+    }
+
+    /// Verifies that `removingListMarkerss` returns a new String that does not contain our List Marker substrings (Attachment).
```
[the remark this answers is not in the document]
Comment (jleandroperez) [reply SimplenoteTests/NSStringSimplenoteTests.swift]: Ahhh thank you!!

**Comment by jleandroperez** [comment]
@aerych comments addressed, thank you sir! (And test plan updated... 🤦‍♂  CMD Z indeed!)

Reeeadyforanotherlook.

Thanks in advance!! 🔥

**Comment by jleandroperez** [comment]
Two bugs spotted, pending fix:

### Bug A:
CMD + Z in Scenario A isn't bringing back the list.

### Bug B:
1. New document
2. Add a bunch of lines
3. Click anywhere in the middle
4. Insert a checklist

As a result the cursor ends up ... not in the current line, but below.

Thank you @aerych !!!

**Comment by jleandroperez** [comment]
@aerych It should (all) be good now.

- **Scenario B**: Turns out the test details were wrong. Updated the text to reflect this
- **Scenrario C**: I've added a new scenario to verify the Cursor's Location bug

And new unit tests were added!! Again, thanks a lot for your time sir ❤️

File: Simplenote/NSString+Simplenote.swift [hunk 2]
```diff
@@ -45,3 +72,43 @@ extension NSString {
...
+            }
+        }
+
+        return output
+    }
+
+    /// Returns a new AttributedString instance, by removing all of the List Markers (with or without trailing spaces) in the receiver
+    ///
+    var removingListMarkers: NSAttributedString {
+        let output = replacingOccurrences(of: .attachmentString + .space, with: String())
```
Comment (aerych) [inline Simplenote/NSString+Simplenote.swift]: Should `.attachmentString + .space` be `.richListMarker` ?

Comment (jleandroperez) [reply Simplenote/NSString+Simplenote.swift]: Nice catch!! Thank you!!

**Pull request merged**